### PR TITLE
PG16 compatibility - Rework PlannedStmt and Query's Permission Info

### PR DESCRIPTION
**Pull request by naisila**
The main issue lies in the following entries of PlannedStmt: {
   rtable
   permInfos
}

Each rtable has an int perminfoindex, and its actual permission info is obtained through the following:
permInfos[perminfoindex]

We had a crash because perminfoindexes were not updated in the finalized planned statement after distributed planner hook.

So, basically, everywhere we set a query's or planned statement's rtable entry, we need to set the rteperminfos/permInfos accordingly.

Relevant PG commits:
https://github.com/postgres/postgres/commit/a61b1f74823c9c4f79c95226a461f1e7a367764b a61b1f74823c9c4f79c95226a461f1e7a367764b
https://github.com/postgres/postgres/commit/b803b7d132e3505ab77c29acf91f3d1caa298f95 b803b7d132e3505ab77c29acf91f3d1caa298f95

### CODE REVIEW

File: src/backend/distributed/planner/distributed_planner.c [hunk 1]
```diff
@@ -1467,6 +1470,29 @@ FinalizeNonRouterPlan(PlannedStmt *localPlan, DistributedPlan *distributedPlan,
...
 	finalPlan->rtable = list_concat(finalPlan->rtable, localPlan->rtable);
 
+#if PG_VERSION_NUM >= PG_VERSION_16
+
+	/*
+	 * Original range table list is concatenated to final plan's range table list
+	 * therefore all the perminfoindexes should be updated to their value
+	 * PLUS the length of final plan's perminfos.
+	 */
+	int list_length_final_permInfos = list_length(finalPlan->permInfos);
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:1480]: I think I noted in the other PR as well, the variable name `list_length_final_permInfos` is confusing as there are lots of lists, and final list maybe confusing.

maybe: final_plan_last_perminfo_index or such could be better?

File: src/backend/distributed/planner/query_colocation_checker.c [hunk 2]
```diff
@@ -83,7 +83,16 @@ CreateColocatedJoinChecker(Query *subquery, PlannerRestrictionContext *restricti
 		 * functions (i.e., FilterPlannerRestrictionForQuery()) rely on queries
 		 * not relations.
 		 */
+#if PG_VERSION_NUM >= PG_VERSION_16
+		RTEPermissionInfo *perminfo = NULL;
+		if (anchorRangeTblEntry->perminfoindex)
+		{
+			perminfo = getRTEPermissionInfo(subquery->rteperminfos, anchorRangeTblEntry);
+		}
+		anchorSubquery = WrapRteRelationIntoSubquery(anchorRangeTblEntry, NIL, perminfo);
```
Comment (onderkalaci) [inline src/backend/distributed/planner/query_colocation_checker.c:92]: we might need WrapRteRelationIntoSubqueryCompat?

File: src/backend/distributed/planner/recursive_planning.c [hunk 3]
```diff
@@ -2392,6 +2445,9 @@ BuildReadIntermediateResultsQuery(List *targetEntryList, List *columnAliasList,
 	Query *resultQuery = makeNode(Query);
 	resultQuery->commandType = CMD_SELECT;
 	resultQuery->rtable = list_make1(rangeTableEntry);
+#if PG_VERSION_NUM >= PG_VERSION_16
```
Comment (onderkalaci) [inline src/backend/distributed/planner/recursive_planning.c:2448]: should we set perminfoindex = 0 at some point in this function?

File: src/include/distributed/recursive_planning.h [hunk 4]
```diff
@@ -42,7 +42,12 @@ extern bool GeneratingSubplans(void);
 extern bool ContainsLocalTableDistributedTableJoin(List *rangeTableList);
 extern void ReplaceRTERelationWithRteSubquery(RangeTblEntry *rangeTableEntry,
 											  List *requiredAttrNumbers,
+#if PG_VERSION_NUM >= PG_VERSION_16
```
Comment (onderkalaci) [inline src/include/distributed/recursive_planning.h:45]: Probably better to have compat functions instead of idfdes in the signature?

File: src/backend/distributed/planner/distributed_planner.c [hunk 5]
```diff
@@ -1504,6 +1530,16 @@ FinalizeRouterPlan(PlannedStmt *localPlan, CustomScan *customScan)
...
 	routerPlan->rtable = list_concat(routerPlan->rtable, localPlan->rtable);
 
+#if PG_VERSION_NUM >= PG_VERSION_16
+
+	/*
+	 * We know that extra remoteScanRangeTableEntry has perminfoindex 0
+	 * therefore we can simply use the perminfos we had in localplan
+	 */
+	Assert(remoteScanRangeTableEntry->perminfoindex == 0);
+	routerPlan->permInfos = localPlan->permInfos;
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:1540]: I'm confused here. Why don't we follow the same logic as `FinalizeNonRouterPlan`? we concat 2 rtables, and I think we should have a single function that does adjustment of the permInfos for the second list. And, we should be able to call that function here?

File: src/backend/distributed/planner/local_distributed_join_planner.c [hunk 6]
```diff
@@ -530,7 +557,13 @@ RequiredAttrNumbersForRelationInternal(Query *queryToProcess, int rteIndex)
  */
 static ConversionCandidates *
 CreateConversionCandidates(PlannerRestrictionContext *plannerRestrictionContext,
```
Comment (onderkalaci) [inline src/backend/distributed/planner/local_distributed_join_planner.c:559]: let's try to avoid ifdefs in the function signatures, they seem super hard to follow

File: src/backend/distributed/planner/merge_planner.c [hunk 7]
```diff
@@ -824,6 +830,20 @@ ConvertRelationRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte,
 	RangeTblEntry *newRangeTableEntry = copyObject(sourceRte);
 	sourceResultsQuery->rtable = list_make1(newRangeTableEntry);
 
+#if PG_VERSION_NUM >= PG_VERSION_16
+	sourceResultsQuery->rteperminfos = NIL;
+	if (newRangeTableEntry->perminfoindex)
+	{
+		/* create permission info for newRangeTableEntry */
+		RTEPermissionInfo *perminfo = getRTEPermissionInfo(mergeQuery->rteperminfos,
+														   newRangeTableEntry);
```
Comment (onderkalaci) [inline src/backend/distributed/planner/merge_planner.c:839]: for completeness, lets use sourceRte as the second param here.

File: src/backend/distributed/planner/merge_planner.c [hunk 8]
```diff
@@ -824,6 +830,20 @@ ConvertRelationRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte,
...
+
+		/* update the subquery's rteperminfos accordingly */
+		newRangeTableEntry->perminfoindex = 1;
```
Comment (onderkalaci) [inline src/backend/distributed/planner/merge_planner.c:842]: The comment says `/* update the subquery's rteperminfos accordingly */`. Wouldn't subqueries always have = 0? I think this is a releation

File: src/backend/distributed/planner/merge_planner.c [hunk 9]
```diff
@@ -849,6 +869,15 @@ ConvertRelationRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte,
 
 	/* replace the function with the constructed subquery */
 	sourceRte->rtekind = RTE_SUBQUERY;
+#if PG_VERSION_NUM >= PG_VERSION_16
+	sourceRte->perminfoindex = 0;
+
+	/*
+	 * Note: we don't need to remove replaced sourceRte from mergeQuery->rteperminfos to avoid
```
Comment (onderkalaci) [inline src/backend/distributed/planner/merge_planner.c:876]: How is this comment accurate? We are in the planner, and ExecCheckPermissions is on the executor, so we are definitely not gone through ExecCheckPermissions.

Do we really need these comments? They seem like development artifacts to me.

File: src/backend/distributed/planner/recursive_planning.c [hunk 10]
```diff
@@ -1778,6 +1804,15 @@ ReplaceRTERelationWithRteSubquery(RangeTblEntry *rangeTableEntry,
 
 	/* replace the function with the constructed subquery */
 	rangeTableEntry->rtekind = RTE_SUBQUERY;
+#if PG_VERSION_NUM >= PG_VERSION_16
+
+	/*
+	 * subquery already contains a copy of this rangeTableEntry's permission info
```
Comment (onderkalaci) [inline src/backend/distributed/planner/recursive_planning.c:1810]: aren't subqueries always have  index=0? So, this comment seems redundant to me.

**Review by onderkalaci — COMMENTED** [review]
I feel like we had to follow test-driven development here, which seems probably a good thing.

Still, do you mind spending maybe 30 mins in the Citus code base to see if there are any places we might need this?

Also, I remember we had some code regarding perm info for COPY, but I guess we are leaving those out as this is more around the `Query *`?

File: src/backend/distributed/planner/distributed_planner.c (hunk 1 above)
Comment (naisila) [reply src/backend/distributed/planner/distributed_planner.c:1480]: Yeah, you did. However I changed something here: previously I was concatenating perminfos right after this line. But now, I concatenate perminfos after the loop - this way it's readable.

However, I agree that a name that includes perminfoindex make this even more readable. So, will change it. Thanks.

File: src/backend/distributed/planner/recursive_planning.c (hunk 3 above)
Comment (naisila) [reply src/backend/distributed/planner/recursive_planning.c:2448]: No need, because rangeTableEntry has been created using makeNode, which means its perminfoindex already has default value of 0.

File: src/backend/distributed/planner/distributed_planner.c (hunk 5 above)
Comment (naisila) [reply src/backend/distributed/planner/distributed_planner.c:1540]: It's true that we concat two rtable lists, but in this case the first list contains just one element (the remoteScanRangeTableEntry), for which we are sure that its perminfoindex hasn't been set - meaning it's 0. So, nothing further needed here.

In FinalizeNonRouterPlan however, we have no idea what the two rtables contain so we just update everything accordingly.

File: src/backend/distributed/planner/merge_planner.c (hunk 8 above)
Comment (naisila) [reply src/backend/distributed/planner/merge_planner.c:842]: I meant to write sourceResultsQuery instead of subquery, thanks.

File: src/backend/distributed/planner/query_colocation_checker.c (hunk 2 above)
Comment (naisila) [reply src/backend/distributed/planner/query_colocation_checker.c:92]: I don't know how that would go.
However, I am thinking of defining RTEPermissionInfo prePG16 and passing it as NULL/NIL in these functions.

**Comment by naisila** [comment]
> Also, I remember we had some code regarding perm info for COPY, but I guess we are leaving those out as this is more around the Query *?

@onderkalaci I couldn't quite get what you mean here. Could you explain this more to me?

File: src/backend/distributed/planner/query_colocation_checker.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/query_colocation_checker.c:92]: sounds good

File: src/backend/distributed/planner/recursive_planning.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/recursive_planning.c:2430]: mostly for readability, but this is also fine

**Comment by onderkalaci** [comment]
> Also, I remember we had some code regarding perm info for COPY, but I guess we are leaving those out as this is more > >around the Query *?
> @onderkalaci I couldn't quite get what you mean here. Could you explain this more to me?


I meant the function CheckCopyPermissions, which it seems we already did in #7005, so you can ignore my comment

**Comment by naisila** [comment]
@onderkalaci 
As per our discussion, because we send the modified query to the standard planner again, the rteperminfos and rtable are adjusted accordingly in the following function:
`add_rte_to_flat_rtable`
Note that this function has the following part:
```C
/*
	 * Add a copy of the RTEPermissionInfo, if any, corresponding to this RTE
	 * to the flattened global list.
	 */
	if (rte->perminfoindex > 0)
	{
		RTEPermissionInfo *perminfo;
		RTEPermissionInfo *newperminfo;

		/* Get the existing one from this query's rteperminfos. */
		perminfo = getRTEPermissionInfo(rteperminfos, newrte);

		/*
		 * Add a new one to finalrteperminfos and copy the contents of the
		 * existing one into it.  Note that addRTEPermissionInfo() also
		 * updates newrte->perminfoindex to point to newperminfo in
		 * finalrteperminfos.
		 */
		newrte->perminfoindex = 0;	/* expected by addRTEPermissionInfo() */
		newperminfo = addRTEPermissionInfo(&glob->finalrteperminfos, newrte);
		memcpy(newperminfo, perminfo, sizeof(RTEPermissionInfo));
	}
```
Because this function needs to get the existing rteperminfo (`/* Get the existing one from this query's rteperminfos. */`), we need to make necessary adjustments throughout the code. So, this explains why we need those adjustments, but we don't need to update rteperminfos in case we have changed the perminfoindex to 0 for some rtables.

File: src/backend/distributed/planner/distributed_planner.c [hunk 11]
```diff
@@ -1081,6 +1084,11 @@ CreateDistributedPlan(uint64 planId, bool allowRecursivePlanning, Query *origina
 	/*
 	 * Plan subqueries and CTEs that cannot be pushed down by recursively
 	 * calling the planner and return the resulting plans to subPlanList.
+	 * Note that GenerateSubplansForSubqueriesAndCTEs will reset perminfoindexes
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:1087]: great comment! Can you add reference to pg function name as well?

File: src/backend/distributed/planner/distributed_planner.c (hunk 5 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/distributed_planner.c:1561]: I feel like in the future we'll need this kind of function, and it'd be hard for anyone to know these details so well.

So, I don't feel strongly about this, but sounds pretty useful to me, so please re-consider.

File: src/include/pg_version_compat.h [hunk 12]
```diff
@@ -144,6 +144,9 @@ object_aclcheck(Oid classid, Oid objectid, Oid roleid, AclMode mode)
 
 typedef bool TU_UpdateIndexes;
 
+/* dummy definition - variable never used but always passed as NULL/NIL */
```
Comment (onderkalaci) [inline src/include/pg_version_compat.h:147]: how is this useful? I couldn't follow actually
